src/storage/src/wal.rs

```diff
@@ -0,0 +1,385 @@
+    /// TODO: this number is not accurate across restarts.
+    total_bytes_written: usize,
+}
+
+impl WriteAheadLog {
+    fn create(id: GlobalId, base_path: PathBuf) -> Result<Self, anyhow::Error> {
+        // First lets create the directory where these log segments will live.
+        // We expect that this directory should not exist and will error if it
+        // does.
+        fs::create_dir(&base_path).with_context(|| {
+            format!(
+                "trying to create wal directory for relation: {} path: {}",
+                id,
+                base_path.display(),
+            )
+        })?;
+
+        // TODO: clean this up
+        let file_path = base_path.join("log-0");
+        let current_file = create_log_segment(&file_path)?;
+
+        let mut ret = Self {
+            id,
+            base_path,
+            current_file,
+            current_sequence_number: 0,
+            current_bytes_written: 0,
+            total_bytes_written: 0,
+        };
+
+        // Write a timestamp message here so we can establish `lower` for the first log
+        // segment.
+        ret.write_progress(0)?;
+        Ok(ret)
+    }
+
+    fn write_inner(&mut self, buf: &[u8]) -> Result<(), anyhow::Error> {
+        let len = buf.len();
+        self.current_file.write_all(&buf).with_context(|| {
+            format!(
+                "failed to write to relation: {} wal segment {}",
+                self.id, self.current_sequence_number
+            )
+        })?;
+        self.current_file.flush().with_context(|| {
+            format!(
+                "failed to flush write to relation: {} wal segment {}",
+                self.id, self.current_sequence_number
+            )
+        })?;
+        self.current_bytes_written += len;
+        self.total_bytes_written += len;
+
+        Ok(())
+    }
+
+    /// Write a set of (Row, Timestamp, Diff) updates to the write-ahead log.
+    fn write(&mut self, updates: &[Update]) -> Result<(), anyhow::Error> {
+        // TODO: The allocation discipline for writes could probably be a lot
+        // better. Each WAL writer could keep a buffer that it writes into
+        // for everything and fall back to allocating stuff if we run out of
+        // space.
+        let mut buf = Vec::new();
+        for update in updates {
+            encode_update(&update.row, update.timestamp, update.diff, &mut buf)?;
+        }
+
+        self.write_inner(&buf)
+    }
+
+    fn write_progress_inner(&mut self, timestamp: Timestamp) -> Result<(), anyhow::Error> {
+        let mut buf = Vec::new();
+        encode_progress(timestamp, &mut buf)?;
+        self.write_inner(&buf)
+    }
+
+    /// Write a progress (alternatively, timestamp closure) message to the write-ahead
+    /// log.
+    ///
+    /// Optionally also switch to a new log segment file if the current log segment file
+    /// grows too large.
+    fn write_progress(&mut self, timestamp: Timestamp) -> Result<(), anyhow::Error> {
```

Comment:
12 bytes are written per second, per table. I must admit I can not envison an actual problem, it just feels unnatural.

Reply:
:D I agree its a bit unnatural. I think _maybe_ you could avoid this if you had the knowledge of "am I closing a timestamp with no updates" in which case you could decide to ignore sending this progress message. But I need to think through it further to decide if that actually makes sense. I'll definitely add a comment documenting this as future work

Reply:
I documented this as a current limitation and todo item